Implement Formatting DataFrame

bryanuribe opened the pull request:
**What this PR does / why we need it:**
This PR implements functionality to format data for the worldmapBehavior and heatmap overlay. Specifically, this PR enables formatting geohash and coordinate based data into an OpenLayer type, Point. 

This PR has three functions:

- **dataFrameToPoints** receives a dataframe object, fieldMappingOptions, and QueryType. The **dataframe** holds the user queried data, which will be translated into an array of Points for the overlay. Furthermore, **fieldMappingOptions and QueryType** hold the data field mapping information. In particular fieldMapping will include user defined mapping about the location based fields, and queryType will include the format that the location data is in.
- **decodeGeoohash** decodes an inputted geohash into coordinates (latitude/longitude).
- **refineInterval** is a helper function used in decodeGeohash.

**Special notes for your reviewer:**
This PR needs to be merged before #16

bryanuribe committed:
Added data formatting for circle overlays

bryanuribe committed:
Merge remote-tracking branch 'origin' into spatial-ol

bryanuribe committed:
Merge branch 'spatial-ol' into format-data-frame

bryanuribe committed:
Merge branch 'spatial-ol' of https://github.com/open-o11y/grafana into spatial-ol

dprokop committed:
TimeSeries panel: Allow adding annotations from the panel (#36220)

* First stab on UI for adding annotations in time series panel

* Extend panel context with annotations api

* Annotations editor UI & CRUD

* Prevent annotation markers to overflow uPlot canvas

* Do not overflow graphing area with region annotations

* Align annotation id type

* Fix exemplar markers positioning

* Use clipping region rather than adjusting annotation region bounds

* Smaller icons

* Improve annotation tooltip and editor auto positioning, reorg code

* Renames

* Enable annotations ctx menu only when adding annotations is allowed

* Wrap setSelect hooks diring init hook

* Use TagFilter instead of TagsInput

* Add id to annotation events

* Add support for cmd+click for adding point annotations

Co-authored-by: Torkel Ödegaard <torkel@grafana.com>

kaydelaney committed:
Timeseries Panel: Retain alerts when migrating from old graph (#36514)

Closes #36106

ying-jeanne committed:
update backend-sdk to v0.106.0 (#36543)

wbrowne committed:
allow for appropriate content-type to be set (#36545)

ying-jeanne committed:
Graphite: Convert to use grafana-plugin-sdk-go contracts (#35798)

* Use Dataframes and extract tags from response

* Fix timestamp conversion

* Add tests for data frame conversion

* Add missing RefID and simplify returning an error

* draft dataframe/sdk convertion for graphite

* intermedia

* modify init because registration failed

* Allocate memory for each data point value

* Remove redundant memory aliasing

* Remove redundant new line

* Sort imports

* Simplify returning nil values

* fix lint

* remove unused jsondata

* add checks on query length

* remove basic auth from request

Co-authored-by: Piotr Jamróz <pm.jamroz@gmail.com>

thisisobate committed:
Docs: added an accessibility section in the pull request guide (#36454)

* Docs: added an accessibility section in the pull request guide

* added more points to the accessibility guideline

* fixes some grammar nits

* changed aria to capital case since its an acronym

* added a reference to the Grafana theme

* improved the wordings for better context

joshhunt committed:
Grafana-UI: Refactor Checkbox to fix alignment issues (#36164)

* Grafana-UI: Fix Checkbox vertical layout issues

* reorganise css

* WIP for fixing checkboxes in manage dashboards

* Cleanup checkbox and fix issue in Dashboard manage page

* update tests

zserge committed:
Gradually remove Macaron web framework (#36325)

* add macaron code to the code base

* remove unused secure cookies support from macaron

* clean up modules

* remove com dependency

* fix silly typos

* little cleanup, remove recovery middleware

* remove logger middleware

* remove static handler and remove unused context methods

* bring inject into macaron codebase

* remove unused applicator

* add back macaron license

* more cleanups in macaron code

* remove unused injector Set method

* remove unused context methods: param to int conversion, body helper type, cookie helpers

* remove action from context

* remove complex environment handling, we only use Env variable

* restore ReplaceAllParams to fix the tests

ifrost committed:
Datasources: Improve error handling for testing data sources (#35120)

* Improve error handling for error messages

The error message will be read from error object from the following properties in the following order:
- message
- data.message
- statusText

* Convert api/ds/query errors to TestingStatus

SQL datasources (mysql, mssql, postgres) and CloudWatch use api/ds/query to test the data source, but previously didn't handle errors returned by this endpoint. If the error cannot be handled it's re-thrown to be handled in public/app/features/datasources/state/actions.ts

* Use async/await instead of Promises

* Remove incorrect type import

TestingStatus is in app/types. Should be pulled down to grafana/data but it depends on HealthCheckResultDetails that is public and lives in grafana/runtime. Ideally TestingStatus should live in grafana/data but I'm not sure if HealthCheckResultDetails can be move there too (?)

* Update packages/grafana-data/src/types/datasource.ts

Co-authored-by: Erik Sundell <erik.sundell@grafana.com>

* Handle errors with no details in toTestingStatus instead of re-throwing

* Update packages/grafana-data/src/types/datasource.ts

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

Co-authored-by: Erik Sundell <erik.sundell@grafana.com>
Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

codesome committed:
Alerting: Allow space in label and annotation names (#36549)

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

joshhunt committed:
AzureMonitor: Resource Picker alignment/spacing tweaks (#35125)

grafanabot committed:
ReleaseNotes: Updated changelog and release notes for 8.0.5 (#36554)

* ReleaseNotes: Updated changelog and release notes for 8.0.5

* Update _index.md

* Update CHANGELOG.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

Co-authored-by: Tobias Skarhed <1438972+tskarhed@users.noreply.github.com>
Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

mckn committed:
Plugins: converted the plugins admin app to a core feature in grafana (#36026)

* moved the plugins admin to core and used the plugins toggle to decide which version to use.

* reverted change.

* changed so the library tab is the default one.

* fixing navigation.
#

* fixed so we have the proper header.

* including the core plugins

* fixed so we display logos for local plugins.

* fixed so we have a working version of plugin catalog.

* removed console logs.

* reverted changes.

* fixing failed test.

tskarhed committed:
Chore: Bump version in latest.json (#36568)

bryanuribe committed:
Merge branch 'spatial-ol' into format-data-frame

bryanuribe committed:
remove debugging statements

bryanuribe committed:
remove local development env variables

bryanuribe committed:
updated comments

bryanuribe committed:
fixed formatting

bryanuribe committed:
Change config variable type declaration

bryanuribe committed:
Revert config type declaration

FZambia committed:
live: better overview in docs (#36506)

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

bryanuribe committed:
Separate queryFormat

bryanuribe committed:
Adjust data formatting process

robbierolin commented on public/app/plugins/panel/geomap/layers/data/utils.ts:
```diff
@@ -0,0 +1,83 @@
+ /**
+   * Function that formats dataframe into a Point[]
+   */
+export function dataFrameToPoints(frame: DataFrame, config: any): Point[] {
```
Can `config` have a more specific type?

bryanuribe replied:
Config is layer specific and can't have a more specific type. However, I've changed the input into the parts the function needs, so that the inputs can be more specific.

cinaglia committed:
Alerting: fix typo in delete dashboard modal prompt (#36270)

mjseaman committed:
Docs: Document which data sources work with caching (#36561)

* Document which data sources work with caching

* add Prometheus and Loki to exclusions

* update list of built-in data sources

I checked in Grafana v8.0.4 and these were the built-in data sources that do work with query caching.

* Apply suggestions from code review

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

* alphabetize list

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

dprokop committed:
User analytics: Add Rudderstack integration (#36567)

* Replace analytics service with Echo backend

* Add Rudderstack integration and general pageview and interaction Echo events

* Update conf/defaults.ini

Co-authored-by: Dan Cech <dcech@grafana.com>

* Update packages/grafana-runtime/src/types/analytics.ts

Co-authored-by: Dan Cech <dcech@grafana.com>

* Update conf/defaults.ini

Co-authored-by: Dan Cech <dcech@grafana.com>

* Update tests

* Force cla check

Co-authored-by: Dan Cech <dcech@grafana.com>

tskarhed committed:
QueryVariable: Default to first option when none is set (#36521)

* TemplateVariables: Default to first option when none is set

* Add test

* Add tests and rewrite logic

aocenas committed:
CloudWatch/Logs: Fix log alerts in new unified alerting (#36558)

* Pass FromAlert header from new alerting

* Add better error messages

AgnesToulet committed:
Config: remove unused Enterprise feature toggles (#36527)

peterholmberg committed:
grafana/ui: Date time picker (#36477)

* styling

* useCallback for onAction

* add flex to accomodate seconds

* fix positioning of the calender

* move input to its own component with state

* wrap callbacks in usecallback

* fix states and add mdx

* add docs

* add tests

* styling fixes for smaller screens

* make date optional

* add test for the changing the input

* more position fixes

* fix an issue with removing the date

* do not show invalid date in input

* more pr feedback

andresmgot committed:
Add AWS/AmazonMQ dimensions (#36573)

ryantxu committed:
Testdata: generate data that can link to USA state codes (#36537)

RichiH committed:
Chore: Add newline at EOF (#36595)

Signed-off-by: Richard Hartmann <richih@richih.org>

gabor committed:
InfluxDB: Flux: fix backward compatibility for some queries (#36603)

* influxdb: flux: better backward-compatibility

* added comment-explanation

ryantxu committed:
Geomap: add initial openlayers alpha panel (#36188)

leeoniya committed:
Stat: use shared data min/max for y auto-ranging (#36497)

marefr committed:
Database: Fix incorrect format of isolation level configuration parameter for MySQL (#36565)

bryanuribe committed:
Changed function input

euniceek commented on public/app/plugins/panel/geomap/layers/data/utils.ts:
```diff
@@ -0,0 +1,84 @@
+
+      // Coordinate Data
+      if (queryFormat.locationType === "coordinates") {
+        lng = row[fieldMapping.latitudeField];
+        lat = row[fieldMapping.longitudeField];
```
```suggestion
        lng = row[fieldMapping.longitudeField];
        lat = row[fieldMapping.latitudeField];
```

euniceek commented on public/app/plugins/panel/geomap/layers/data/utils.ts:
```diff
@@ -0,0 +1,84 @@
+      else if (queryFormat.locationType === "geohash"){
+        const encodedGeohash = row[fieldMapping.geohashField];
+        const decodedGeohash = decodeGeohash(encodedGeohash);
+        lng = decodedGeohash.latitude;
+        lat = decodedGeohash.longitude;
```
```suggestion
        lng = decodedGeohash.longitude;
        lat = decodedGeohash.latitude;
```

bryanuribe committed:
Fixed longitude and latitude mapping

robbierolin reviewed (approved):
LGTM

torkelo committed:
Theme: Update theme style guide and docs with new theme info and variable picking guide (#36530)

* Theme: Update theme style guide and docs with new theme info and variable picking guide

* Minor spell fix

FZambia committed:
live: better error logging in push API (#36601)

torkelo committed:
DashboardList: Fix issue not re-fetching dashboard list after variable change (#36591)

torkelo committed:
Select: Minor style and usability improvements (#36628)

* Select: Minor style & hover improvements

* added testid

* change to aria label

codesome committed:
Alerting: Fix potential panic in Alertmanager when starting up (#36562)

* Alerting: Fix potential panic in Alertmanager when starting up

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

* Fix reviews

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

gabor committed:
influxdb: influxql: better tag-value filtering (#36570)

pricek committed:
Table: Add minimum column width field for auto resizing (#35351)

* TablePanel: Adds minimum column width field

The table panel uses virtual rendering of rows, which prevents resizing
based on the contents of a column. A minimum column width field
simulates this ability.

* Fixed indentation issue

* Updates the table panel documentation

Added the "Minimum column width" option to table panel docs and updated
the information for the "Column width" option in the docs.

mckn committed:
Plugins: display enterprise plugins in the plugins catalog (#36599)

* displaying enterprise plugins in the list.

* added place holder for tests and removed unused code.

* added test for the browse page.

* added empty test file.

* added some more tests.

tskarhed committed:
A11y: ARIA hide image/link duplicate in news panel (#36642)

* fix(a11y): Hide image/link duplicate

* fix: proper heading and time markup

torkelo committed:
Links: Fixes issue with some links causing full page reload (#36631)

zoltanbedi committed:
Tracing: add way to configure trace to logs start and end time (#34995)

* Tracing: add way to configure trace to logs start and end time

* Use the span's time by default

* Update docs

* Update time inputs to use strings instead of number

* Support negative values as well

* Add info about using negative value

* Don't round up Loki range

* Update docs/sources/datasources/jaeger.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

* Wording for doc

* Round adjusted start and end time

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

zoltanbedi committed:
Tempo: show hex strings instead of uints for ids (#36471)

brentcetinich committed:
Add StreamName dimension for AWS/KinesisVideo namespace (#36655)

codesome committed:
Alerting: A better and cleaner way to know if Alertmanager is initialised (#36659)

Signed-off-by: Ganesh Vernekar <ganeshvern@gmail.com>

ivanahuckova committed:
Explore: Refactor QueryRow to use ConnectedProps (#36604)

* Refactor QueryRow

* Upddate import in test

* Simplify typing

* Add copy when loading instance

tskarhed committed:
BottomNavLinks: Accessible text for image and icon (#36650)

torkelo committed:
Transformations: State feature (alpha/beta) and more (#36630)

* Adding plugin state feature to transforms

* initial help box

* New HelpBox component

* More progress

* Testing

* Removing HelpBox, simple new design, new active state for OperationRowAction

* Updated tests

* Fixed typing issue

* Removed AlphaNotice

* Made focus and enter key trigger OnClick and sorted transformations

* Fixed e2e tests

ashharrison90 committed:
TimeSeries: Improve tooltip positioning when tooltip overflows (#36440)

* TimeSeries: Improve tooltip positioning when tooltip overflows

* VizTooltip: Use react-popper, extract positioning calculation into util function + add unit tests

* VizTooltip: Keep ref as tooltipRef

* Use popper only for VizTooltip positioning

* VizTooltip: Set altAxis to true to prevent overflow on y axis

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

ivanahuckova committed:
Explore: Fix running of log queries with shift-enter shortcut (#36600)

* Re-run query if we are running query fromo query row

* Fix imports in tests

* Change solution

* Dispatch clear cache

grobinson-grafana committed:
Add ValueString to the documentation for alerts (#36654)

mckn committed:
Plugins: adding possibility to install enterprise plugins if you have a valid license. (#36653)

domasx2 committed:
Alerting: better detect cortex/loki ruler api (#36030)

* wip

* beter detect non existing rules stuff

* fix useIsRuleEditable

* test for detecting editable-ness of a rules datasource

* tests!

* fix lint errors

bryanuribe committed:
Start of unit tests

domasx2 committed:
Alerting: add toggle in Loki/Prom datasource config to opt out of alerting ui (#36552)

* wip

* fix useIsRuleEditable

* test for detecting editable-ness of a rules datasource

* tests!

* fix lint errors

* add alerting ui opt-out toggle to loki and prom datasources

* Apply suggestions from code review

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

euniceek committed:
GeoMap: Implement circle overlays (#36680)

tharun208 committed:
Password Field Improvements (#36160)

* Password: added show password functionality
added autcomplete props
created password component

Fixes #28721

* addressed review changes and added unit tests

* wrapped passwordField component in forwardRef

* fix validation and tests

idafurjes committed:
Fix Postgres query handling null values for smallint (#36648)

* Fix Postgres query handling null values for smallint

* Fix converting to int16

torkelo committed:
Transformations: Enable / disable toggle for transformation rows  (#36668)

* Disable transform feature

* Change icon

* Added more clear way to show it's disabled

glindstedt committed:
Avoid breaking on fieldConfig without defaults field (#36666)

This would result in a `Dashboard init failed` error when migrating
dashboards with a folded panel that has a `fieldConfig` but has not
defined `fieldConfig.defaults`.

wbrowne committed:
Plugins: Improve grafana-cli UX + API response messaging for plugin install incompatibility scenario (#36556)

* improve UX for plugin install incompatability

* refactor test

torkelo committed:
Visualizations: Dynamically set any config (min, max, unit, color, thresholds) from query results (#36548)

* initial steps for config from data

* Moving to core and separate transforms

* Progress

* Rows to fields are starting to work

* Config from query transform working

* UI progress

* More scenarios working

* Update public/app/core/components/TransformersUI/rowsToFields/rowsToFields.ts

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

* transform all

* Refactor

* UI starting to work

* Add matcher UI to transform

* Apply to self

* Adding a reducer option

* Value mapping via new all values reducer

* value mappings workg add -A

* Minor changes

* Improving UI and adding test dashboards

* RowsToFieldsTransformerEditor tests

* Added tests for FieldToConfigMapping Editor

* Added test for ConfigFromQueryTransformerEditor

* Minor UI tweaks

* Added missing test

* Added label extraction

* unified mapping

* Progress refactoring

* Updates

* UI tweaks

* Rename

* Updates

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

axelavargas committed:
Annotations: Add custom label for custom values in builtin annotations (#36691)

perplexityjeff committed:
Documentation: Update image_rendering.md to include libgbm1 (#35850)

* Update image_rendering.md

* Updated with feedback 

Updated with feedback

kevisong committed:
Legacy Alerting: Replace simplejson with a struct in webhook notification channel (#34952)

* Add type for webhook body

* Update pkg/services/alerting/notifiers/webhook.go

Co-authored-by: Sofia Papagiannaki <papagian@users.noreply.github.com>

olbo98 committed:
Explore: page title indicates which datasource is being used  (#36699)

* Explore: quick fix for page title

* Explore: update page title in componentDidUpdate with info about the current datasources

* Explore: change tests to incorporate the new page title

* Explore: remove updatePageDocumentTitle since the recent changes replace it

jackw committed:
Chore: Sync lock file with package.json (#36667)

davidmparrott committed:
change template expansion missing value handling (#36679)

sarahzinger committed:
Upgrade node version for e2e tests (#36620)

aknuds1 committed:
grafana-plugin-ci-e2e: Fix README; bump version (#36716)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

FZambia committed:
Live: handle influx input with incomplete/asymmetrical field set (#36664)

jackw committed:
Plugins: Don't show uninstall button for core plugins in catalog (#36671)

* fix(catalog): don't show un/install buttons if core plugin

* test(catalog): add PluginDetails test case for core plugins

ryantxu committed:
DataLinks: support 'onClick' from table panels (#36682)

davidmparrott committed:
Alerting: annotation on state change (#36535)

* WIP

* Add annotation on alert state change

* move annotation creation to manager

* praise the linter!

* add debug msg when creating annotation

bryanuribe committed:
Add unit tests for data formatting

bryanuribe committed:
Updated comments for data formatter

bryanuribe committed:
Fixed formatting

ryantxu committed:
Geomap: rename "circles" layer to "markers" layer (#36727)

bryanuribe committed:
Merge branch 'master' into format-data-frame

bryanuribe committed:
Merge master
